### PR TITLE
refactor: rebuild lsp command by api

### LINKED ISSUE #2068: Add a setup option to configure border of the floating window of LspInfo and others
As per discussion in gitter/matrix, @glepnir asked to create an issue and tag him, and promised to inmpement it later :grin:

### PR DESCRIPTION
**Pull request by glepnir**
cc @justinmk 

Fix #2068

### CODE REVIEW

**Comment by ranjithshegde** [comment]
@glepnir #1984 already includes some of these changes

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:13]: Can you move it to the top? So defining `g:lspconfig` will disable it entirely.

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:2]: Can `require 'lspconfig'` be inlined in all places where `lspconfig` is used? Then we don't force the module to be loaded.

Comment (glepnir) [reply plugin/lspconfig.lua:2]: done

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:40]:
```suggestion
    return vim.tbl_filter(function(s)
      return s:sub(1, #arg) == arg
    end, require 'lspconfig'.available_servers())
```

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:32]:
```suggestion
  require 'lspconfig'.lsp_start(args.args)
```

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:49]:
```suggestion
  for _, client in ipairs(require 'lspconfig'.util.get_clients_from_cmd_args(cmd_args)) do
    client.stop()
  end
```

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:58]:
```suggestion
  require 'lspconfig'.lsp_restart(cmd_args)
```

**Comment by justinmk** [comment]
let's revisit after https://github.com/neovim/nvim-lspconfig/pull/1984 ? Though maybe this isn't needed if `nvim_create_user_command` serves the purpose well enough? 

What's the use case that isn't covered by `nvim_create_user_command` ?

File: plugin/lspconfig.lua
Comment (glepnir) [reply plugin/lspconfig.lua]: done

**Comment by glepnir** [comment]
This is different from #1984. I just replaced lspconfig.vim. and  fixed #2068  now we can use `LspInfo border=single` to set the border . why not provide a config to set the border. because I think the `lspconfig` just provide the server config. compare with `vim.cmd` use string format to generate the `autocmd` use api will be more better:) and  neovim use lua file in everywhere now :)

**Review by justinmk — APPROVED** [review]
yeah, this looks like good changes.

File: plugin/lspconfig.lua
Comment (ii14) [reply plugin/lspconfig.lua:31]: @glepnir you've missed this one

Comment (glepnir) [reply plugin/lspconfig.lua]: oh sry. fix

**Comment by ranjithshegde** [comment]
This conflicts with #1984 because `lspconfig.lua` logic introduced in that PR is incompatible. All these `user_commands` are moved to `plugin/lspconfig.lua`, with `lua/lspconfig.lua` just creatin/assigning a metatable for lsp_configurations.

I would recommend please wait till #1984 is merged. then this PR is simply modifying the `border` key and make it accessible by `fargs/nargs`

**Comment by glepnir** [comment]
@ranjithshegde  because you don't reply my review so I continue work on this . maybe this will be merge or not @ii14  what do you think?

**Comment by ranjithshegde** [comment]
> @ranjithshegde because you don't reply my review so I continue work on this . maybe this will be merge or not @ii14 what do you think?

My apologies. For whatever reason I wasn't subscribed to this thread. 

If you could please wait until that is merged, then the job of this PR will be very very simple. All the logic will already have been introduced in #1984,
 then you will simply have to make the `border` key modifiable by user.

**Comment by glepnir** [comment]
> now we can use LspInfo border=single to set the border . why not provide a config to set the border. because I think the lspconfig just provide the server config.

this project only provide the server config. not provide any other config should be better. so pass the argument into the cmdline is my way. not use a `border` key config .

**Comment by ii14** [comment]
> what do you think?

I don't know, I'm only reviewing the code :)

**Pull request closed without merging**
